ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryMonitor.java

```diff
@@ -120,32 +123,52 @@ private void maybeRestartQueries() {
     final long now = ticker.read();
 
     // Restart queries that has passed the waiting timeout
-    final List<QueryId> deleteRetryEvents = new ArrayList<>();
-    queriesRetries.entrySet().stream()
-        .forEach(mapEntry -> {
-          final QueryId queryId = mapEntry.getKey();
-          final RetryEvent retryEvent = mapEntry.getValue();
-          final Optional<PersistentQueryMetadata> query = ksqlEngine.getPersistentQuery(queryId);
-
-          // Query was terminated manually if no present
-          if (!query.isPresent()) {
-            deleteRetryEvents.add(queryId);
-          } else if (query.get().isError() && now > retryEvent.nextRestartTimeMs()) {
-            // Retry again if it's still in ERROR state
-            retryEvent.restart();
-          } else if (now > retryEvent.queryHealthyTime()) {
-            // Clean the errors queue & delete the query from future retries now the query is
-            // healthy
-            query.ifPresent(QueryMetadata::clearErrors);
-            deleteRetryEvents.add(queryId);
-          }
-        });
-
-    deleteRetryEvents.stream().forEach(queriesRetries::remove);
+    final Iterator<Map.Entry<QueryId, RetryEvent>> it = queriesRetries.entrySet().iterator();
+    while (it.hasNext()) {
+      final Map.Entry<QueryId, RetryEvent> mapEntry = it.next();
+
+      final QueryId queryId = mapEntry.getKey();
+      final RetryEvent retryEvent = mapEntry.getValue();
+      final Optional<PersistentQueryMetadata> query = ksqlEngine.getPersistentQuery(queryId);
+
+      // Query was terminated manually if no present
+      if (!query.isPresent()) {
+        it.remove();
+      } else {
+        final PersistentQueryMetadata queryMetadata = query.get();
+        switch (queryMetadata.getState()) {
+          case ERROR:
+            if (now >= retryEvent.nextRestartTimeMs()) {
+              retryEvent.restart(queryMetadata);
+            }
+            break;
+          case RUNNING:
+          case REBALANCING:
+            if (now >= retryEvent.queryHealthyTime()) {
+              // Clean the errors queue & delete the query from future retries now the query is
+              // healthy and has been running after some threshold time
+              query.ifPresent(QueryMetadata::clearErrors);
```

Comment:
nit: we know it's present, can we just `queryMetadata.clearErrors()`?

Reply:
Done